ddl/column.go

```diff
@@ -856,16 +947,29 @@ func replaceOldIndexes(tblInfo *model.TableInfo, changingIdxs []*model.IndexInfo
 	}
 }
 
-func updateNewIndexesCols(tblInfo *model.TableInfo, oldName, newName model.CIStr, newOffset int) {
-	for _, idx := range tblInfo.Indices {
+// updateNewIndexesCols updates the changing indexes column name&offset, and
+// filter out the indexes that can be removed.
+func updateNewIndexesCols(changingIdxs []*model.IndexInfo,
+	oldName, newName model.CIStr, newOffset int) []*model.IndexInfo {
+	indexesToRemove := make([]*model.IndexInfo, 0, len(changingIdxs))
+	for _, idx := range changingIdxs {
+		var hasChangingCol bool
 		for i, col := range idx.Columns {
 			if col.Name.L == oldName.L {
 				idx.Columns[i].Name = newName
 				idx.Columns[i].Offset = newOffset
-				break
+			} else {
+				if !hasChangingCol {
+					hasChangingCol = len(getChangingColumnOriginName(col)) != len(col.Name.L)
+				}
```

Comment:
```suggestion
				if getChangingColumnOriginName(col) != col.Name.L {
									hasChangingCol = true
									break
				}
```

And one question here, why only compare `len`?

Reply:
Different length means it is a changing column.

Reply:
I know that, but it looks a little bit strange. I think compare name directly is better.

Reply:
OK.